Adds helm chart for heptio/ark

domcar opened the pull request:
## Premise
Helm cannot handle properly CRD becauses it has a validation mechanism that checks the installation before the CRD are actually created,
hence each resource that uses a CRD cannot be validated because the CRD doesn't exist yet!

## Solution
The solution here is to create CRD via helm chart, and only after (using a `post-install`) to install the resources with a container.
The container has the only job to execute a `kubectl create -f filename` and create the resources.

At the same time the resources created with the hook are completely transparent to Helm, that is, when you delete the
chart those resources remain there. Hence we need a second hook for deleting them

domcar committed:
first commit

domcar committed:
changes SA name

domcar committed:
adds correct rbac rules

domcar committed:
renames files; add more customizable vars

domcar committed:
updates readme

domcar committed:
adds notes

domcar committed:
removes config values

domcar committed:
changes email in chart

domcar commented:
CLA signed

domcar committed:
updates readme

domcar committed:
test changes author

domcar committed:
test change email

domcar committed:
test change email

domcar committed:
adds prerequisites in readme

unguiculus commented:
/assign

domcar committed:
fixes typo

domcar commented:
/assign @unguiculus

domcar committed:
adds AWS

domcar committed:
updates to version 0.7

domcar committed:
updates version in chart

domcar committed:
adds repo source; removes unnecessary values

unguiculus commented on incubator/ark-server/README.md:
```diff
@@ -0,0 +1,46 @@
+## ConfigMap customization
+Since we want to have a customizable chart it's important that the configmap is a template and not a static file.
+To do this we add the keyword `tpl` when reading the file
+- {{ (tpl (.Files.Glob "static/*").AsConfig .) | indent 2 }}
```
Why don't you just add it to the templates folder? It doesn't make sense to me to have it in a `static` folder when it is not static.

domcar replied:
For the reason I explained in the Premise `Helm cannot handle properly CRD becauses it has a validation mechanism that checks the installation before the CRD are actually created,
hence each resource that uses a CRD cannot be validated because the CRD doesn't exist yet!`.

If you move the file in `templates` you get the error `Error: apiVersion "ark.heptio.com/v1" in ark-server/templates/01-config-deploy.yaml is not available`.

But I agree that the name `static` is not appropriate

unguiculus commented on incubator/ark-server/README.md:
```diff
@@ -0,0 +1,46 @@
+
+### Heptio Secret
+Ark server needs a IAM service accoutn in order to run, if you don't have it you must create it: 
+https://github.com/heptio/ark/blob/v0.6.0/docs/cloud-provider-specifics.md#gcp
```
0.7.0

unguiculus commented on incubator/ark-server/README.md:
```diff
@@ -0,0 +1,46 @@
+
+### Bucket and Project name
+Please change bucket and project/region name in the values.yaml file
+See here for possible values: https://github.com/heptio/ark/blob/v0.6.0/docs/cloud-provider-specifics.md
```
0.7.0

unguiculus commented on incubator/ark-server/static/01-config-deploy.yaml:
```diff
@@ -0,0 +1,62 @@
+
+---
+apiVersion: apps/v1beta1
+kind: Deployment
```
Why is this not in `templates`?

domcar replied:
I moved it to `templates`

unguiculus reviewed (changes requested):
Thanks for the PR. Before I continue reviewing it, please make sure you follow our review guidelines and apply best practices. In general we want separate files per resources (with an exception for RBAC stuff).

https://github.com/kubernetes/charts/blob/master/REVIEW_GUIDELINES.md
https://github.com/kubernetes/helm/blob/master/docs/chart_best_practices/rbac.md

domcar committed:
moves deployment to templates

domcar committed:
renames folder

domcar committed:
updaates to 0.7.1

domcar committed:
creates ark sa in helpers; separates files according to object type

domcar committed:
updates version in chart

domcar committed:
adds home to chart

ncdc commented on incubator/ark-server/templates/deployment.yaml:
```diff
@@ -0,0 +1,35 @@
+apiVersion: apps/v1beta1
+kind: Deployment
+metadata:
+  namespace: {{ .Values.namespace.name }}
```
You'll want this to be a namespace that is separate from where the backups/restores/etc are placed. In Ark v0.7.1, this is `heptio-ark-server` by default.

ncdc commented on incubator/ark-server/templates/deployment.yaml:
```diff
@@ -0,0 +1,35 @@
+          command:
+            - /ark
+          args:
+            - server
```
You'll want to add
```
- --namespace
- {{ <a variable for the namespace in which backups, schedules, restores, and the Ark config go> }}
```

So you'll probably need 1 variable for the namespace where the server lives, and a second variable for the namespace where the backups etc go.

domcar replied:
`- {{ <a variable for the namespace in which backups, schedules, restores, and the Ark config go> }}`

You wrote also  `ark config`, but I read the new yaml files and it remeains in heptio-ark-server namespace, right?

ncdc replied:
No, the config remains in `heptio-ark` along with backups, schedules, and restores.

ncdc commented:
@domcar in Ark v0.7.1, we are now installing the Ark server pod in a different namespace from the backups/restores/etc. Please see https://github.com/heptio/ark/pull/322 for the changes to our yaml files. I'd recommend doing the same here.

domcar commented:
@ncdc thanks, I'll add the chagnes to the yaml files

domcar committed:
updates to v0.7.1

unguiculus commented on incubator/ark-server/templates/crd.yaml:
```diff
@@ -0,0 +1,93 @@
+    plural: backups
+    kind: Backup
+
+---
```
Please create one file per resource (except for RBAC).

domcar replied:
No problem doing this, but can you explain me the reason? I thought some logical grouping can help understand better how it works

unguiculus commented on incubator/ark-server/README.md:
```diff
@@ -0,0 +1,46 @@
+- `configmap.yaml` Configmap will be mounted to the hook container as a file and subsequently used as k8s manifest for deploy or deletion
+
+## ConfigMap customization
+Since we want to have a customizable chart it's important that the configmap is a template and not a static file.
```
Then why don't you put it into the templates folder directly?

domcar replied:
I answered you in another comment:

For the reason I explained in the Premise `Helm cannot handle properly CRD becauses it has a validation mechanism that checks the installation before the CRD are actually created,
hence each resource that uses a CRD cannot be validated because the CRD doesn't exist yet!`.

If you move the file in `templates` you get the error `Error: apiVersion "ark.heptio.com/v1" in ark-server/templates/01-config-deploy.yaml is not available`.

But I agree that the name `static` is not appropriate

unguiculus commented on incubator/ark-server/configuration/01-config-deploy.yaml:
```diff
@@ -0,0 +1,25 @@
+  name: {{ required "PLEASE SET VALUE" .Values.configuration.cloudprovider  }}
+  config:
+    {{- if eq .Values.configuration.cloudprovider "gcp" }}
+    project: {{ required "PLEASE SET VALUE" .Values.configuration.project  }}
```
Unfortunaly, you can't use required values when there is no default in `values.yaml`. CI will fail because there is currently no way to provide defaults in the CI job.

domcar replied:
Yes I understand, I'll point out in the Readme that those values must be changed before roll out.

I used the `required` keyword because there are not default values that can be used for project or bucket, they are specific for each setup.

unguiculus commented on incubator/ark-server/templates/crd.yaml:
```diff
@@ -0,0 +1,93 @@
+metadata:
+  name: backups.ark.heptio.com
+  labels:
+    component: ark
```
The chart doesn't have multiple components. I would just remove this label everywhere.

domcar replied:
ok, I just had left it as it was in the original files

unguiculus commented on incubator/ark-server/templates/deployment.yaml:
```diff
@@ -0,0 +1,37 @@
+metadata:
+  namespace: {{ .Values.namespace.server.name }}
+  name: ark
+spec:
```
Needs `spec.selector`

unguiculus commented on incubator/ark-server/templates/deployment.yaml:
```diff
@@ -0,0 +1,37 @@
+apiVersion: apps/v1beta1
```
`apps/v1beta2`

We officially support the current and previous minor versions of Kubernetes.

unguiculus commented on incubator/ark-server/templates/deployment.yaml:
```diff
@@ -0,0 +1,37 @@
+kind: Deployment
+metadata:
+  namespace: {{ .Values.namespace.server.name }}
+  name: ark
```
Use fullname template for the name.

unguiculus commented on incubator/ark-server/templates/deployment.yaml:
```diff
@@ -0,0 +1,37 @@
+  template:
+    metadata:
+      labels:
+        component: ark
```
`app` and `release` labels should go here.

unguiculus commented on incubator/ark-server/templates/rbac_ark.yaml:
```diff
@@ -0,0 +1,37 @@
+{{- if .Values.serviceAccount.server.create }}
+apiVersion: v1
+kind: ServiceAccount
```
Put service account into its own file.

unguiculus commented on incubator/ark-server/values.yaml:
```diff
@@ -0,0 +1,49 @@
+serviceAccount:
+  hook:
+    create: true
+    name: deploy-ark
```
Default should be empty

unguiculus commented on incubator/ark-server/values.yaml:
```diff
@@ -0,0 +1,49 @@
+    name: deploy-ark
+  server:
+    create: true
+    name: ark
```
Default should be empty

unguiculus reviewed (changes requested):
I only did a quick initial review and gave you a few pointers. Note that namepace handling can be problematic. Helm implicitly creates the name space you specify with `--namespace` if it does not exist. If you then have a template for this namespace in your chart, its creation will fail. I don't like the fact the Ark now uses two namespaces, which is not your fault, though. Seems like an unfortunate workaround. Maybe it's the only way around these deletion issues which, to be honest, I had already experienced myself ;).

domcar committed:
modifies chart according to best practices

domcar committed:
removes blank line

domcar committed:
adds delete backup hook to solve issue crd backup not deleting

unguiculus commented on incubator/ark-server/templates/deployment.yaml:
```diff
@@ -0,0 +1,42 @@
+  replicas: 1
+  selector:
+    matchLabels:
+      component: ark
```
Selector should have `app` and `release` labels.

unguiculus commented on incubator/ark-server/templates/deployment.yaml:
```diff
@@ -0,0 +1,42 @@
+  template:
+    metadata:
+      labels:
+        component: ark
```
Selector should have `app` and `release` labels.

unguiculus commented on incubator/ark-server/templates/hook_delete.yaml:
```diff
@@ -0,0 +1,28 @@
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: delete-ark-config
```
Make list indentation consistent. This should indent one more level. Apply everywhere.

unguiculus reviewed (changes requested):
Sorry for the delay. You had asked about the split into multiple files. I think it is better when you see right away what's in the file. You may put the CRDs into a subfolder, though. That'd probably be best.

Wherever possible, please base resource names on the `fullname` template.

It's common practice here to use hyphens in filenames vs. underscores. Would you mind following that practice, please?

I'd suggest you move the chart to stable before we merge it.

domcar committed:
adjusts indentation, renames file with using hyphene

domcar committed:
moves folder to stable

domcar commented:
I modified according to your last comments

ncdc commented:
FYI @domcar we are working on changes to Ark to return back to a single namespace for both the server & the backups, so you might want to hold off until we have that published.

pierreozoux commented on stable/ark-server/templates/rbac.yaml:
```diff
@@ -0,0 +1,57 @@
+  apiGroup: rbac.authorization.k8s.io
+{{- end }}
+
+{{- if .Values.rbac.create }}
```
Why do you have 2 ifs?

domcar replied:
you are right, it's a mistake, i'll change it

domcar committed:
remove unnecessary test condition

rtluckie commented:
@ncdc any updates?

ncdc commented:
We plan to release Ark 0.8 next week, consolidating back to 1 namespace

On Wed, Apr 11, 2018 at 6:28 PM Ryan Luckie <notifications@github.com>
wrote:

> @ncdc <https://github.com/ncdc> any updates?
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/charts/pull/3795#issuecomment-380615807>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAABYqNwY8wWUVDSiOJ2PdRFl6pnsm7rks5tnoN2gaJpZM4SNc-C>
> .
>

gianrubio commented on stable/ark-server/Chart.yaml:
```diff
@@ -0,0 +1,11 @@
+apiVersion: v1
+appVersion: "1.0"
```
appVersion should be 0.7.1

jody-frankowski replied:
Indeed, and I would add that `version` should be at `1.0.0`, because it's the first version of this chart.

ncdc replied:
We'll be releasing Ark v0.8.0 later this week. We'll be back to a single namespace instead of two. I'd recommend waiting until then before merging this chart if you can!

gianrubio commented on stable/ark-server/templates/deployment.yaml:
```diff
@@ -0,0 +1,46 @@
+      app: {{ template "ark-server.name" . }}
+  template:
+    metadata:
+      labels:
```
please add annotation here so it'll be easy to use ark with kube2iam

```
    {{- if .Values.podAnnotations }}
      annotations:
{{ toYaml .Values.podAnnotations | indent 8 }}
    {{- end }}
```

gianrubio commented on stable/ark-server/templates/hook-delete-backups.yaml:
```diff
@@ -0,0 +1,29 @@
+      restartPolicy: Never
+      containers:
+        - name: delete-ark-backups
+          image: claranet/gcloud-kubectl-docker
```
I'm a bit scared to run a kubectl image from untrusted source, could you add this as a variable

```image: {{ .Values.kubectl.image.repository }}/{{ .Values.kubectl.image.tag }}```

domcar replied:
I understand, although it's just a simple image with those app installed

jody-frankowski replied:
I agree with @gianrubio. Moreover, as is the image doesn't point to a precise version, and that is against the [best practices](https://github.com/kubernetes/helm/blob/master/docs/chart_best_practices/pods.md#images). Lastly I wasn't able to find sources for this image on [docker hub](https://hub.docker.com/r/claranet/gcloud-kubectl-docker/) or on [github](https://github.com/claranet) :)

domcar replied:
I already changed it in the last commit :)

gianrubio commented on stable/ark-server/templates/hook-delete.yaml:
```diff
@@ -0,0 +1,29 @@
+      restartPolicy: Never
+      containers:
+        - name: delete-ark-config
+          image: claranet/gcloud-kubectl-docker
```
same as above

gianrubio commented on stable/ark-server/README.md:
````diff
@@ -0,0 +1,51 @@
+```
+
+### Configuration
+PLease change thevalues.yaml according to your setup
````
You missed a table with key/values, similar to [this](https://github.com/kubernetes/charts/tree/master/stable/buildkite#configuration)

nrb commented:
Version v0.8.0, which came out last week, introduced a new CRD, https://github.com/heptio/ark/blob/v0.8.0/examples/common/00-prereqs.yaml#L91.

gianrubio commented:
@domcar are you working on this? If you don't mind I can pick your commits and make the required changes in a new PR

domcar commented:
@gianrubio I was planning on working on it starting this week. Are you in a hurry or can you wait?

gianrubio commented:
I can wait, thanks @domcar

domcar committed:
Update to v0.8.1; Add support for Azure

domcar commented:
@ncdc Hello, I updated the chart to 0.8.1 Can you please have a look in case I missed something? I also added azure as cloudprovider, could you please test it on azure since I don't have an account there?

ncdc commented:
@skriss @nrb ptal

ncdc commented on stable/ark-server/README.md:
```diff
@@ -0,0 +1,63 @@
+
+## Prerequisites
+
+### Heptio Secret
```
Maybe say "Secret for cloud provider credentials"?

ncdc commented on stable/ark-server/README.md:
```diff
@@ -0,0 +1,63 @@
+
+### Heptio Secret
+Ark server needs a IAM service account in order to run, if you don't have it you must create it.
+This is the guide for gcp: https://github.com/heptio/ark/blob/v0.8.1/docs/gcp-config.md#create-service-account
```
I'd recommend pointing to https://heptio.github.io/ark/v0.8.1/cloud-common as a starting point that contains links to configure Ark for each cloud provider.

ncdc commented on stable/ark-server/README.md:
```diff
@@ -0,0 +1,63 @@
+
+### Configuration
+Please change the values.yaml according to your setup
+See here for the official documentation https://github.com/heptio/ark/blob/v0.8.1/docs/config-definition.md
```
https://heptio.github.io/ark/v0.8.1/config-definition

ncdc commented on stable/ark-server/README.md:
````diff
@@ -0,0 +1,63 @@
+```
+
+## Remove heptio/ark
+Rememebr that when you remove ark all backups remain untouched
````
Remember

gianrubio commented on stable/ark-server/values.yaml:
```diff
@@ -0,0 +1,51 @@
+
+podAnnotations:
+
+kubectl:
```
Did you leave empty on purpose?

domcar replied:
Yes. 
Do you think it's better if i add the string ` iam.amazonaws.com/role: arn:aws:iam::<AWS_ACCOUNT_ID>:role/<HEPTIO_ARK_ROLE_NAME>`
and take those 2 vars from values.yaml ?

domcar replied:
I added the string, I thinkg it's better

jody-frankowski replied:
I think he was talking about the kubectl variable.
By the charts best practices, default values should either be enough to install and run the chart or produce a warning that lists the variables the user needs to fill.
Moreover the annotation you added is AWS specific, and not everyone run their cluster on aws (gke, aks, on-premise, ...).

domcar replied:
ah, I misunderstood, i'll change it, tnx. 
Regarding the annotation, It's only aws because as far as I saw in the ark docs, it can be used only with aws: https://github.com/heptio/ark/blob/v0.8.1/examples/aws/10-deployment-kube2iam.yaml

jody-frankowski replied:
About the annotation: Since it depends on where your cluster is installed AND it depends on whether you want to use kube2iam, I would just comment it, with another comment explaining what needs to be done to use kube2iam.

About the kubectl variable: The repository should be `claranet/gcloud-kubectl-docker` and the tag should be a proper tag according to https://github.com/kubernetes/helm/blob/master/docs/chart_best_practices/pods.md#images . Lastly the docker image should be open source according to https://github.com/kubernetes/charts/blob/master/CONTRIBUTING.md#technical-requirements

domcar committed:
Update readme

domcar committed:
Add annotation for kube2iam

domcar committed:
Add image for hooks. Update readme

sortigoza commented:
If I run this more than 1 time, 

Create ark secret and  install chart, delete if exist.
~~~
set -e

kubectl get namespace $ARK_NAMESPACE || kubectl create namespace $ARK_NAMESPACE
kubectl delete secret cloud-credentials \
  --namespace $ARK_NAMESPACE || true
kubectl create secret generic cloud-credentials \
  --namespace $ARK_NAMESPACE \
  --from-file cloud=credentials-ark

helm del --purge ark-server || true
helm install local/ark-server \
  --name ark-server \
  --namespace $ARK_NAMESPACE \
  -f values-ark-server.yaml
~~~

I get the following error:

~~~
secret "cloud-credentials" deleted
secret "cloud-credentials" created
Error: jobs.batch "delete-ark-config" already exists <<<<
Error: a release named ark-server already exists. <<<<
Run: helm ls --all ark-server; to check the status of the release
Or run: helm del --purge ark-server; to delete it
~~~

~~~
HELM version
Client: &version.Version{SemVer:"v2.9.0", GitCommit:"f6025bb9ee7daf9fee0026541c90a6f557a3e0bc", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.9.0", GitCommit:"f6025bb9ee7daf9fee0026541c90a6f557a3e0bc", GitTreeState:"clean"}
KUBERNETES version
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.1", GitCommit:"d4ab47518836c750f9949b9e0d387f20fb92260b", GitTreeState:"clean", BuildDate:"2018-04-13T22:27:55Z", GoVersion:"go1.9.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.7", GitCommit:"dd5e1a2978fd0b97d9b78e1564398aeea7e7fe92", GitTreeState:"clean", BuildDate:"2018-04-18T23:58:35Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
~~~

This has to do with:

~~~
command: ["kubectl","delete","-f","/tmp/"]
Error from server (NotFound): error when deleting "/tmp/01-config-deploy.yaml": configs.ark.heptio.com "default" not found
~~~

If the configmap was never created or failed to be created. The hook-deploy job will fail and we won't be able to del the deployed chart.  We may need additional validation logic here.

domcar commented:
@sortigoza I repeated the same steps you listed but I didn't encounter your problem.
What did you change in your `values.yaml`?
I guess you set to false the `namespace create` right? Something else?

unguiculus commented on stable/ark-server/templates/namespace.yaml:
```diff
@@ -0,0 +1,6 @@
+{{- if .Values.namespace.create }}
```
Remove the namespace resource. This is unnecessary. Helm creates a namespace automatically if it doesn't exist. Thus, installation will fail if the release namespace is the same as `{{ .Values.namespace }}`. The user can specify the namespace they want using the CLI. That's all we need.

domcar replied:
I think this could lead to a problem with `rbac.yaml` because there I have to specify a namespace

jody-frankowski replied:
Use `Release.Namespace` https://docs.helm.sh/chart_template_guide/#built-in-objects

domcar replied:
Thanks

unguiculus commented on stable/ark-server/templates/_helpers.tpl:
```diff
@@ -0,0 +1,54 @@
+{{/*
+Create the name of the service account to use for creating or deleting the ark config
+*/}}
+{{- define "ark-server.hookSA" -}}
```
Please rename to `ark-server.hookServiceAccount`

unguiculus commented on stable/ark-server/templates/_helpers.tpl:
```diff
@@ -0,0 +1,54 @@
+Create the name of the service account to use for creating or deleting the ark server
+*/}}
+{{- define "ark-server.sa" -}}
+{{- if .Values.serviceAccount.server.create -}}
```
Please rename to `ark-server.serverServiceAccount`

unguiculus commented on stable/ark-server/templates/backups.yaml:
```diff
@@ -0,0 +1,16 @@
+metadata:
+  name: backups.ark.heptio.com
+  labels:
+    chart: {{ .Chart.Name }}-{{ .Chart.Version }}
```
Use `chart: {{ template "ark-server.chart" . }}` throughout the chart.

unguiculus commented on stable/ark-server/templates/hook-delete.yaml:
```diff
@@ -0,0 +1,29 @@
+        - name: delete-ark-config
+          image: {{ required "A docker image with kubectl" .Values.kubectl.image.repository }}/{{ required "A docker image with kubectl" .Values.kubectl.image.tag }}
+          imagePullPolicy: Always
+          command: ["kubectl","delete","-f","/tmp/"]
```
Add spaces after commas: `["kubectl", "delete", "-f", "/tmp/"]`

unguiculus commented on stable/ark-server/templates/hook-deploy.yaml:
```diff
@@ -0,0 +1,28 @@
+        - name: deploy-ark-config
+          image: {{ required "A docker image with kubectl" .Values.kubectl.image.repository }}/{{ required "A docker image with kubectl" .Values.kubectl.image.tag }}
+          imagePullPolicy: Always
+          command: ["kubectl","create","-f","/tmp/"]
```
Add spaces after commas: `["kubectl", "create", "-f", "/tmp/"]`

unguiculus commented on stable/ark-server/templates/hook-delete.yaml:
```diff
@@ -0,0 +1,29 @@
+          command: ["kubectl","delete","-f","/tmp/"]
+          volumeMounts:
+            - name: ark-server-config
+              mountPath: /tmp
```
I'd give the directory a dedicated name instead of using `/tmp`.

domcar replied:
I don't think it's very important, in the end the two hooks are temporary, they should deploy and die, so I though the /tmp was ok

unguiculus commented on stable/ark-server/templates/configmap.yaml:
```diff
@@ -0,0 +1,7 @@
+  name: ark-server-config
+  namespace: {{ .Values.namespace.name }}
+data:
+{{ (tpl (.Files.Glob "configuration/*").AsConfig .) | indent 2 }}
```
There is only one file. Use `Files.Get` instead.

unguiculus replied:
If you use the alternative approach I [suggested](https://github.com/kubernetes/charts/pull/3795/files/00d250eb194d9d6c920553dc151f4c0e374a6c49#diff-4f20687d4680b53414b33ce4051b4a56), you could just do this:

```yaml
data:
  default:
{{ toYaml .Values.configuration | indent 4 }}
```

domcar replied:
This is a nice suggestion, tnx

domcar replied:
I think there is a problem with this approach because I have to specify the namespace:
```
configuration:
  apiVersion: ark.heptio.com/v1
  kind: Config
  metadata:
    namespace: heptio-ark
```
And I can't do that in `values.yaml`
What do you think?

unguiculus commented on stable/ark-server/templates/deployment.yaml:
```diff
@@ -0,0 +1,56 @@
+        {{- end }}
+        - name: plugins
+          emptyDir: {}
+      {{- if eq .Values.configuration.cloudprovider "azure" }}
```
I would just make the `nodeSelector` configurable without any Azure-specific condition. You may also want to add `tolerations`.

domcar replied:
I know, but I would like the chart to follow the ark deploy and they use the `nodeselector` for azure only

jody-frankowski replied:
I agree with you since you have managed to squeeze the 3 different provider-specific deployments of Ark into one.
IMHO, you should provide a way to customize the `nodeSelector` while keeping azure's check (and tolerations too as @unguiculus suggests).

unguiculus commented on stable/ark-server/configuration/01-config-deploy.yaml:
```diff
@@ -0,0 +1,26 @@
+apiVersion: ark.heptio.com/v1
+kind: Config
```
How about making the Config directly configurable via `values.yaml`? This will not tie you to a specific set of config options. Otherwise, a new chart release would be required if a new Ark release adds new options.

unguiculus commented on stable/ark-server/README.md:
```diff
@@ -0,0 +1,65 @@
+
+## Premise
+Helm cannot handle properly CRD becauses it has a validation mechanism that checks the installation before the CRD are actually created,
+hence each resource that uses a CRD cannot be validated because the CRD doesn't exist yet!
```
My suggestion:

In general, Helm cannot install CRDs and resources based on these CRDs in the same Helm chart because CRDs need to be installed before CRD resources can be created and Helm cannot guarantee the correct ordering for this to work.

As a workaround, the chart creates a `Config` resource via `post-install` hook. Since resources created by hooks are not managed by Helm, a `pre-delete` hook removes the `Config` CRD when the release is deleted.

domcar replied:
ok

unguiculus commented on stable/ark-server/README.md:
```diff
@@ -0,0 +1,65 @@
+- `configmap.yaml` Configmap will be mounted to the hook container as a file and subsequently used as k8s manifest for deploy or deletion
+
+## ConfigMap customization
+Since we want to have a customizable chart it's important that the configmap is a template and not a static file.
```
the configmap **be** a template

unguiculus commented on stable/ark-server/README.md:
```diff
@@ -0,0 +1,65 @@
+
+## ConfigMap customization
+Since we want to have a customizable chart it's important that the configmap is a template and not a static file.
+To do this we add the keyword `tpl` when reading the file
```
Obsolete with my alternative suggestion.

unguiculus commented on stable/ark-server/README.md:
```diff
@@ -0,0 +1,65 @@
+At the same time the resources created with the hook are completely transparent to Helm, that is, when you delete the
+chart those resources remain there. Hence we need a sencond hook for deleting them (see hook-delete.yaml)
+
+## Content
```
I would remove this section.

unguiculus commented on stable/ark-server/README.md:
```diff
@@ -0,0 +1,65 @@
+Ark server needs a IAM service account in order to run, if you don't have it you must create it.
+Please follow the official documentation: https://heptio.github.io/ark/v0.8.1/cloud-common
+
+Don't forget the step to create the secret
```
It should be possible to have the chart create the secret with the alternative to specify an `existingSecret`.

domcar replied:
At the moment one can only specify an `existingSecret`, if you have a suggestion how Helm can  create it without manual action from the User please let me know, I'd like to add this feature.

unguiculus commented on stable/ark-server/templates/deployment.yaml:
```diff
@@ -0,0 +1,56 @@
+apiVersion: apps/v1beta1
```
I'd suggest you change this to `apps/v1beta2` and add a `spec.selector` below.

domcar replied:
I would like the chart to follow the ark deploy without divergence

unguiculus commented on stable/ark-server/templates/deployment.yaml:
```diff
@@ -0,0 +1,56 @@
+  template:
+    metadata:
+      labels:
+        component: ark
```
I don't think we need a component label since there are no multiple components.

domcar replied:
I would like the chart to follow the ark deploy without divergence

unguiculus commented on stable/ark-server/templates/deployment.yaml:
```diff
@@ -0,0 +1,56 @@
+      serviceAccountName: {{ template "ark-server.sa" . }}
+      containers:
+        - name: ark
+          image: "{{ .Values.image.repository }}:{{ .Values.image.tag }}"
```
Add `imagePullPolicy`

unguiculus reviewed (changes requested):
Please make sure you add standard labels to all resources.

How about renaming the chart to just `ark`?

domcar committed:
Rename serviceaccount

domcar committed:
Use Get instead of Glob

domcar committed:
Remove namespace

domcar committed:
Add PullPolicy; Modify readme

domcar commented:
@unguiculus yes, I'll rename it just `ark`, the reason for `ark-server` was for a private project but here doesn't make sense

domcar committed:
Rename Chart ark

domcar committed:
Add standard labels to resources

domcar committed:
Add customizable tolertion and nodeselector

svyotov commented on stable/ark/templates/deployment.yaml:
```diff
@@ -0,0 +1,59 @@
+kind: Deployment
+metadata:
+  name: {{ template "ark.fullname" . }}
+  release: {{ .Release.Name }}
```
"Error: error validating \"\": error validating data: found invalid field release for v1.ObjectMeta"
I am getting the error above for `release` and for `app`.
Probably they need to be removed?

laghoule replied:
Missing labels. Should be something like:
```
metadata:
  name: {{ template "ark.fullname" . }}
  labels:
    app: {{ template "ark.name" . }}
    chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
    release: {{ .Release.Name }}
    heritage: {{ .Release.Service }}
```

domcar replied:
Thanks, i'll add them now

domcar committed:
Add missing labels; Use image with tag

unguiculus commented on stable/ark/values.yaml:
```diff
@@ -0,0 +1,53 @@
+kubectl:
+  image:
+    repository: claranet
+    tag: gcloud-kubectl-docker:1.0.0
```
```yaml
kubectl:
  image:
    repository: claranet/gcloud-kubectl-docker
    tag: 1.0.0
```

svyotov replied:
Also this is a good alternative image: https://hub.docker.com/r/lachlanevenson/k8s-kubectl/ as far as I can see.

unguiculus commented on stable/ark/values.yaml:
```diff
@@ -0,0 +1,53 @@
+
+# Only kube2iam: change the AWS_ACCOUNT_ID and HEPTIO_ARK_ROLE_NAME
+podAnnotations:
+  iam.amazonaws.com/role: arn:aws:iam::<AWS_ACCOUNT_ID>:role/<HEPTIO_ARK_ROLE_NAME>
```
Should not be there by default. Comment it out.

unguiculus commented on stable/ark/templates/deployment.yaml:
```diff
@@ -0,0 +1,61 @@
+  release: {{ .Release.Name }}
+  app: {{ template "ark.name" . }}
+  chart: {{ template "ark.chart" . }}
+  heritage: {{ .Release.Service }}
```
Labels should go under `labels`.

unguiculus commented on stable/ark/templates/deployment.yaml:
```diff
@@ -0,0 +1,61 @@
+      {{- if eq .Values.configuration.cloudprovider "azure" }}
+      nodeSelector:
+        beta.kubernetes.io/os: linux
+{{ toYaml .Values.nodeSelector | indent 8 }}
```
Node selector should not depend on Azure. 

And as pointed out earlier, I'd remove the special selector for Azure. It is only necessary when you run on Azure and have Windows node. In this case, we can expect the user to configure that.

unguiculus commented on stable/ark/templates/rbac.yaml:
```diff
@@ -0,0 +1,55 @@
+  kind: ClusterRole
+  name: cluster-admin
+  apiGroup: rbac.authorization.k8s.io
+
```
Remove empty line.

unguiculus commented on stable/ark/values.yaml:
```diff
@@ -0,0 +1,53 @@
+  pullPolicy: IfNotPresent
+
+# this must be created manually before helm install, see README
+secret:
```
Why not have an option to create the secret and alternatively configure an `existingSecret`?

domcar replied:
I would rather let the user create their own secrets, it looks to me too much to create this secret from helm

unguiculus commented on stable/ark/values.yaml:
```diff
@@ -0,0 +1,53 @@
+    repository: claranet
+    tag: gcloud-kubectl-docker:1.0.0
+
+tolerations:
```
`tolerations: []`

unguiculus commented on stable/ark/values.yaml:
```diff
@@ -0,0 +1,53 @@
+
+tolerations:
+
+nodeSelector:
```
`nodeSelector: {}`

svyotov commented on stable/ark/templates/hook-delete.yaml:
```diff
@@ -0,0 +1,32 @@
+      restartPolicy: Never
+      containers:
+        - name: delete-ark-config
+          image: {{ required "A docker image with kubectl" .Values.kubectl.image.repository }}/{{ required "A docker image with kubectl" .Values.kubectl.image.tag }}
```
tag and image should be split by `:`
```
{{ required "A docker image with kubectl" .Values.kubectl.image.repository }}:{{ required "A docker image with kubectl" .Values.kubectl.image.tag }}
```
It would be probably be nice to be able to specify kubernetes docker credentials, for those using their own docker registry: 
```
mountPath: /tmp
{{- if .Values.kubectl.image.secret }}
          imagePullSecrets:
              - name: {{ .Values.kubectl.image.secret }}
{{- end }}
```

svyotov commented on stable/ark/templates/deployment.yaml:
```diff
@@ -0,0 +1,61 @@
+            - name: plugins
+              mountPath: /plugins
+          {{- if (or (eq .Values.configuration.cloudprovider "aws") (eq .Values.configuration.cloudprovider "gcp")) }}
+            - name: cloud-credentials
```
The name in my opinion should be a configurable variable.

svyotov commented on stable/ark/templates/hook-delete.yaml:
```diff
@@ -0,0 +1,32 @@
+        - name: delete-ark-config
+          image: {{ required "A docker image with kubectl" .Values.kubectl.image.repository }}/{{ required "A docker image with kubectl" .Values.kubectl.image.tag }}
+          imagePullPolicy: Always
+          command: ["kubectl", "delete", "-f", "/tmp/"]
```
If the image has entry point `kubectl` or command `kubectl`,  you can pass just the arguments, without the `kubectl` such as here: https://hub.docker.com/r/lachlanevenson/k8s-kubectl/ imo.

domcar committed:
Implement suggestion

unguiculus commented:
I'll probably have some time to more thoroughly test and review the PR today. If you don't mind, I'm going to help fix things.

domcar commented:
@unguiculus you're welcome to help :)

unguiculus committed:
Various updates

unguiculus commented:
I made quite a few changes. I'll continue with this next week and hope to push something then.

unguiculus commented:
/ok-to-test

unguiculus commented:
I pushed a few changes:

* Add option to create or use existing secret.
* Create the config resource directly in the configmap without using a separate template.
* Expose the complete config directly via values removing any cloud provider specifics there. The only place where cloud provider specific stuff is still needed is in the deployment. The config resource configuration maps directly now.
* Disable deployment and hooks if no provider and bucket are set in order to make CI happy. Ark won't start up without configuration but we cannot configure it properly for CI.
* Add config table to readme.

I tested it successfully on my GKE cluster.

unguiculus committed:
Merge remote-tracking branch 'upstream/master' into feature/ark

unguiculus committed:
Add missing if block around delete hook

unguiculus commented:
@domcar @svyotov @sortigoza @ncdc PTAL

domcar commented:
@unguiculus I'm fine with the changes and I could successully deploy using an existing secret. But it doesn't work when putting the secrets into `.Values.credentials.secretContents`. Did you successfully test it? If so can you please tell me how you put the secret into the `values.yaml` ?

unguiculus commented:
@domcar I tested it again. It works with limitations. Note that base64 encoding is done by the chart. You need to specify values in plain text.

**A few examples:**

```yaml
credentials:
  secretContents:
    AZURE_SUBSCRIPTION_ID: ...
    AZURE_TENANT_ID: ...
    AZURE_RESOURCE_GROUP: ...
    AZURE_CLIENT_ID: ...
    AZURE_CLIENT_SECRET: ...
    AZURE_STORAGE_ACCOUNT_ID: ...
    AZURE_STORAGE_KEY: ...
```

```
--set credentials.secretContents.AZURE_SUBSCRIPTION_ID=my-id \
--set credentials.secretContents.AZURE_TENANT_ID=my-id
```

```yaml
credentials:
  secretContents:
    cloud: |
      {
        "type": "service_account",
        "project_id": "my-project",
        "private_key_id": "123456789",
        "private_key": "-----BEGIN PRIVATE KEY-----\n...\n-----END PRIVATE KEY-----\n",
        "client_email": "service@my-project.iam.gserviceaccount.com",
        "client_id": "123456789",
        "auth_uri": "https://accounts.google.com/o/oauth2/auth",
        "token_uri": "https://accounts.google.com/o/oauth2/token",
        "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
        "client_x509_cert_url": "https://www.googleapis.com/robot/v1/metadata/x509/service%40my-project.iam.gserviceaccount.com"
      }
```

I didn't manage to pass the GCP service account key from the last example on the commandline. Some escaping would probably be necessary. I got `wrong type for value; expected string; got []interface {}`.

domcar commented:
@unguiculus ok thanks, I managed to deploy on GCP with the secret in the `values.yaml` and it looks like everything is working

unguiculus commented:
@domcar Can we merge it?

domcar commented:
yes :)

unguiculus commented:
/lgtm
